memory/memory/memory.py

```diff
@@ -0,0 +1,120 @@
+import kubernetes_asyncio as kube
+
+from hailtop.config import get_deploy_config
+from hailtop.google_storage import GCS
+from hailtop.hail_logging import AccessLogger
+from hailtop.tls import get_in_cluster_server_ssl_context
+from hailtop.utils import AsyncWorkerPool, retry_transient_errors
+from gear import setup_aiohttp_session, rest_authenticated_users_only
+
+uvloop.install()
+
+BATCH_PODS_NAMESPACE = os.environ['HAIL_BATCH_PODS_NAMESPACE']
+log = logging.getLogger('batch')
+routes = web.RouteTableDef()
+
+socket = '/redis/redis.sock'
+
+
+@routes.get('/healthcheck')
+async def healthcheck(request):  # pylint: disable=unused-argument
+    return web.Response()
+
+
+@routes.get('/api/v1alpha/objects')
+@rest_authenticated_users_only
+async def get_object(request, userdata):
+    filename = request.query.get('q')
+    etag = request.query.get('etag')
+    userinfo = await get_or_add_user(request.app, userdata)
+    username = userdata['username']
+    log.info(f'memory: request for object {filename} from user {username}')
+    result = await get_file_or_none(request.app, username, userinfo, filename, etag)
+    if result is None:
+        raise web.HTTPNotFound()
+    etag, body = result
+    return web.Response(headers={'ETag': etag}, body=body)
+
+
+async def get_or_add_user(app, userdata):
+    users = app['users']
+    username = userdata['username']
+    if username not in users:
+        k8s_client = app['k8s_client']
+        gsa_key_secret = await retry_transient_errors(
+            k8s_client.read_namespaced_secret,
+            userdata['gsa_key_secret_name'],
+            BATCH_PODS_NAMESPACE,
+            _request_timeout=5.0)
+        gsa_key = base64.b64decode(gsa_key_secret.data['key.json']).decode()
+        users[username] = {'fs': GCS(blocking_pool=app['thread_pool'], key=json.loads(gsa_key))}
+    return users[username]
+
+
+def make_redis_key(username, filepath):
+    return f'{ username }_{ filepath }'
+
+
+async def get_file_or_none(app, username, userinfo, filepath, etag):
+    filekey = make_redis_key(username, filepath)
+    fs = userinfo['fs']
+
+    cached_etag, result = await app['redis_pool'].execute('HMGET', filekey, 'etag', 'body')
+    if cached_etag is not None and cached_etag.decode('ascii') == etag:
+        log.info(f"memory: Retrieved file {filepath} for user {username} with etag'{etag}'")
+        return cached_etag.decode('ascii'), result
+
+    log.info(f"memory: Couldn't retrieve file {filepath} for user {username}: current version not in cache.")
+    if filekey not in app['files_in_progress']:
+        try:
+            log.info(f"memory: Loading {filepath} to cache for user {username}")
+            app['worker_pool'].call_nowait(load_file, app['redis_pool'], app['files_in_progress'], filekey, fs, filepath)
+        except asyncio.QueueFull:
+            pass
+    return None
+
+
+async def load_file(redis, files, file_key, fs, filepath):
+    try:
+        files.add(file_key)
```

Comment:
Ah, sorry I missed this:  there should be a test here.  You need to do nothing if you're already loading the file.

Reply:
I realized I hadn't implemented this as intended: the actual semantics I'd wanted here were to only add the file to the queue if it wasn't already in the queue, and not just if it wasn't being actively loaded. I moved the `files.add(file_key)` line to reflect that; I *think* that makes adding a check here redundant, but I've added one just in case.

Reply:
Hmm, I still not sure about the current logic.  I don't see where the loading is skipped if it is queued.  I feel like you want something like, at the call site:

```
  if not file_key in files:
    files.add(file_key)
    call_nowait(load_file, ...)

  def load_file():
    try:
      # load that file
    finally:
      files.remove(file_key)
```